docs/architecture/adr-001-transfer-module-implementation.md

```diff
@@ -0,0 +1,177 @@
+   `transfer` (on the counterparty), it will be received in `recv_packet` as
+   `transfer/channel-0/my_token`. `ibc-rs` strips `transfer/channel-0/` from the
+   denom, so `coin.denom` would be `my_token` for unescrowing.
+2. Validate that the token is native and **not** an IBC-created token by
+   referencing the `minted_token_id_to_name` state.
+    * If the token ID is found in the `minted_token_id_to_name` state, it is an
+      IBC-created token, and the transfer is rejected. This step only fails when
+      the counterparty chain produces a malicious IBC transfer `send_packet()`.
+3. Obtain the escrow address for a specified port and channel pair, similar to
+   the escrowing step.
+4. Verify that the escrow account has a sufficient balance.
+5. Unescrow the token from the escrow account to the receiver's address.
+
+### Minting Tokens - Receiver on Rollup with Sender as Source
+
+1. Obtain the full denom by prefixing it (e.g., `transfer/channel-0/uatom`).
+   `ibc-rs` handles prefixing a base denom with the specified port and channel
+   IDs.
+2. Retrieve the token ID by checking if a token for the given denom has been
+   previously created by the IBC module using the `minted_token_name_to_id` map.
+   * If yes, use that `TokenId`.
+   * If no, [create a new
+     token](https://github.com/informalsystems/sovereign-ibc/blob/4e37dc4bb88624765384d1662549c00e991acc4a/modules/sov-ibc-transfer/src/context.rs#L105)
+     with the name set to the `denom`, obtain the token ID, and store the pair
+     of *token name*, *token ID*, and its flip in the the
+     `minted_token_id_to_name` and `minted_token_name_to_id` state maps.
+   * NOTE: when IBC initiates the creation of a new token, the `IbcTransfer`
+     address is designated as the authorized minter.
+   * NOTE: In this steps we ensure that the `context` object needed for token
+     creation uses the `ibc_transfer` address as the `sender` by constructing a
+     new context object.
+3. Mint tokens to the receiver's address with the specified amount in the
+   `MsgRecvPacket` message.
+
+### Burning Tokens - Sender on Rollup with Receiver as Source
+
+1. Verify that the `memo` field does not exceed maximum allowed length to
+   prevent large memos overwhelming the system. We set the maximum memo length
+   to 32768 bytes like the `ibc-go`.
+2. Obtain the `TokenId` using the denom from the `minted_token_name_to_id` map.
+   * If the token ID is not found, the transfer is rejected.
+3. Confirm that the sender has a sufficient balance.
+4. Burn tokens from the sender's address by calling `burn` on the `bank` module.
+
+Therefore, As a primary rule, when transferring a native token, the token ID is
+used as the denom. For IBC-created tokens, the regular prefixed denomination is
+utilized as the denom. This may pose a challenge for front-ends to correctly
+identify the token type when crafting the related appropriate transfer message.
+Here existing RPC methods can assist.
+
```

Comment:
We should add some sample ics20 transactions (`send_packet`) among (at least) three sovereign chains to avoid confusion.

Reply:
I can do this, and you can verify.

Reply:
Cool! Thanks